server/server.js

```diff
@@ -23,11 +24,189 @@ app.use(express.static(uploadsStaticDir));
 app.use(express.json());
 
 app.get('/api/hello', (req, res) => {
-  res.json({ message: 'Hello World!' });
+  res.json({ message: 'Hello!' });
+});
+
+app.get('/api/public/Tables/Recipes', async (req, res, next) => {
+  try {
+    const sql = `
+      select *
+        from "Recipes";
+    `;
+    const result = await db.query(sql);
+    const recipes = result.rows;
+    res.json(recipes);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.get('/api/public/Tables/Ingredients', async (req, res, next) => {
+  try {
+    const sql = `
+      select *
+        from "Ingredients";
+    `;
+    const result = await db.query(sql);
+    const ingredients = result.rows;
+    res.json(ingredients);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.get('/api/public/Tables/Users', async (req, res, next) => {
+  try {
+    const sql = `
+      select *
+        from "Users";
+    `;
+    const result = await db.query(sql);
+    const users = result.rows;
+    res.json(users);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.get('/api/public/Tables/Favorites', async (req, res, next) => {
+  try {
+    const sql = `
+      select *
+        from "Favorites"
+        join "Recipes" using ("recipeId");
+    `;
+    const result = await db.query(sql);
+    const favorites = result.rows;
+    favorites.forEach((recipe) => {
+      fixIngredients(recipe);
+    });
+    res.json(favorites);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.get('/api/public/Tables/Favorites/:recipeId', async (req, res, next) => {
+  try {
+    const recipeId = Number(req.params.recipeId);
+    if (!Number.isInteger(recipeId) || recipeId <= 0) {
+      throw new ClientError(400, `${recipeId} must be a positive integer`);
+    }
+    const sql = `
+      select *
+        from "Favorites"
+        join "Recipes" using ("recipeId")
+      where "recipeId" = $1;
+    `;
+    const params = [recipeId];
+    const result = await db.query(sql, params);
+    const [recipe] = result.rows;
+    if (!recipe) {
+      throw new ClientError(404, `cannot find recipe with recipeId ${recipeId}`);
+    } else {
+      fixIngredients(recipe);
+      res.json(recipe);
+    }
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/public/Tables/Favorites', async (req, res, next) => {
+  try {
+    const userId = Number(req.body.userId);
+    const recipeId = Number(req.body.recipeId);
+    if (Number.isNaN(recipeId) || !Number.isInteger(recipeId) || recipeId < 0) {
+      throw new ClientError(400, 'recipeId must be a positive integer');
+    }
+    if (Number.isNaN(recipeId) || !Number.isInteger(userId) || userId < 0) {
+      throw new ClientError(400, 'userId must be a positive integer');
+    }
+    if (!userId || !recipeId) {
```

Comment:
You probably want to check this _before_ checking if they are NaN. Or don't check at all, because if they are undefined they will be NaN.